src/libraries/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -925,6 +927,21 @@ private static void EnsureInitialized()
             }
         }
 
+        internal static T LazyInitializeConsoleAndField<T>([NotNull] ref T? field, Func<T> initializer) where T : class =>
+            Volatile.Read(ref field) ?? LazyInitializeConsoleAndFieldCore(ref field, initializer);
+
+        private static T LazyInitializeConsoleAndFieldCore<T>([NotNull] ref T? field, Func<T> initializer) where T : class
+        {
+            // We don't call ConsolePal.EnsureInitialized from within the initializer function to avoid deadlocks.
+            // ConsolePal.EnsureInitialized takes a lock on Console.Out.
+            // Console.EnsureInitialized takes a lock on InternalSyncObject.
+            // When called from the initializer, a deadlock could happen when another thread locks these objects
+            // in the opposite order.
+            EnsureInitialized();
+
+            return Console.EnsureInitialized(ref field, initializer);
+        }
+
```

Comment:
@stephentoub I was thinking to add an assert in `EnsureInitializedCore` that checks the current thread isn't locking `InternalSyncObject`. I think there is no nice way to write it using `Monitor` methods.

Reply:
```C#
Debug.Assert(!Monitor.IsEntered(InternalSyncObject));
```
?

Reply:
`Monitor.IsEntered`... I assumed it was for any thread.
That is exactly what I need. Thanks!